src/parser.rs

```diff
@@ -154,59 +153,158 @@ impl<'a> Parser<'a> {
         }
     }
 
-    /// Tries to parse an expression.
-    fn parse_expression(&mut self) -> Option<Expression> {
+    fn parse_parenthesised_expression(&mut self) -> Option<Expression> {
+        assert_eq!(self.token.kind(), Kind::LeftParen);
         let start = self.position;
-
-        if let Some(subexpr) = self.parse_subexpression() {
-            match self.token.kind() {
-                Kind::Plus => self.parse_binary_expression(start, subexpr, OpName::Plus),
-                Kind::Minus => self.parse_binary_expression(start, subexpr, OpName::Minus),
-                Kind::Divide => self.parse_binary_expression(start, subexpr, OpName::Divide),
-                Kind::Star => self.parse_binary_expression(start, subexpr, OpName::Multiply),
-                Kind::SemiColon
-                | Kind::Comma
-                | Kind::LeftParen
-                | Kind::RightParen
-                | Kind::LeftBrace
-                | Kind::RightBrace => Some(subexpr),
-                _ => {
-                    self.reset(start);
-                    None
-                }
+        self.read_token(); // consume '('
+        if let Some(expression) = self.parse_expression() {
+            if self.token.kind() == Kind::RightParen {
+                self.read_token(); // consume ')'
+                Some(expression)
+            } else {
+                self.reset(start);
+                None
             }
         } else {
             self.reset(start);
             None
         }
     }
 
-    /// Tries to parse a binary subexrpression.
-    fn parse_binary_expression(
+    fn build_nested_expressions(
         &mut self,
-        start: usize,
-        left: Expression,
-        operation: OpName,
+        mut expressions: VecDeque<Expression>,
+        mut operators: VecDeque<OpName>,
     ) -> Option<Expression> {
-        self.read_token();
-        self.parse_expression().map_or_else(
-            || {
+        if expressions.len() != operators.len() + 1 {
+            panic!(
```

Comment:
I can clean up these `panic`s into `Err`/`Result` after merge if you like

Reply:
That would be awesome, thanks!